File: make_table.py
import openpyxl
from openpyxl.styles import Font
import re
import cv2
from PIL import Image
from pdf2image import convert_from_path
import numpy as np
import os
import win32com.client

flag_war_start = False
import_track = []
clan_war_time = 0


def maketable_introduction(flag):
    msg = "集計表を作成するお＾ｑ＾(Cancelで作成中止)\n" \
          "```自チームのメンバー [ふつきん,P2,P3,P4,P5,P6]\n自チームの各得点 [80,80,80,80,80,80]\n\n" \
          "敵チーム名 [ABC]\n対戦回数 [1]\n" \
          "敵チームのメンバー [P7,P8,P9,P10,P11,P12]\n敵チームの各得点 [80,80,80,80,80,80]"

    if flag == 1:
        return msg + "```※コースと交流戦日付は記入不要だお＾ｑ＾"
    else:
        return msg + "\n\nコース名 [T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12]\n"\
                     "自チームの選択コース [1,2,3,4,5,6]\n\n日付 [2020/01/01]```"


def maketable_check_input(str, flag):
    safe, msg = True, ""
    n1, n2 = str.count('['), str.count(']')

    if not str.startswith("自チームのメンバー"):
        safe = False
    elif n1 >= 6 and n2 >= 6:
        input = str.replace("[", "]").split("]")
        for i in range(1, 12, 2):
            if (i != 5 and i != 7) and input[i].count(',') != 5:
                safe = False
                msg = "名前または得点が6つずつ入力されていないかも？＾ｑ＾"
        if flag == 0 and safe:
            if not input[13].count(',') == 11:
                safe = False
            msg = "コースが12個入力されていないかも？＾ｑ＾"
        if flag == 1 and safe:
            if not n1 == 6 and n2 == 6:
                safe = False
            msg = "余計なデータがある or []で閉じられていない箇所があるかも？＾ｑ＾"
        elif flag == 0 and safe:
            if not n1 == 9 and n2 == 9:
                safe = False
            msg = "余計なデータがある or []で閉じられていない箇所があるかも？＾ｑ＾"
    else:
        safe = False
        msg = "[]で閉じられていない箇所があるかも？＾ｑ＾"

    return safe, msg


def make_table(input, flag):
    source_excel_pass = 'clanwar_table.xlsx'
    created_excel_pass = 'created_sheet.xlsx'
    pdf_pass = 'created_sheet.pdf'
    table_image_pass = 'table.png'
    bg_image_pass = 'clanwar_table_bg.png'
    created_image_pass = 'created_sheet.png'

    book = openpyxl.load_workbook(source_excel_pass)  # 集計表を作成
    sheet = book['6 vs 6']
    input = input.replace("[", "]").split("]")

    ally_name = input[1].split(',')
    for i in range(6):
        sheet.cell(3 + i, 4, ally_name[i])

    ally_point = input[3].split(',')
    for i in range(6):
        sheet.cell(3 + i, 6, int(ally_point[i]))

    sheet.cell(9, 1, input[5])

    enemy_name = input[9].split(',')
    for i in range(6):
        sheet.cell(9 + i, 4, enemy_name[i])

    enemy_point = input[11].split(',')
    for i in range(6):
        sheet.cell(9 + i, 6, int(enemy_point[i]))

    if flag == 1:
        for i in range(12):
            if "__" in import_track[i]:
                sheet.cell(3 + i, 12, re.sub('[_*]', "", import_track[i]))
                cell = sheet.cell(row=3 + i, column=12).coordinate
                sheet[cell].font = Font(color='00B0F0', size=26, name='Cataneo BT')
            else:
                sheet.cell(3 + i, 12, import_track[i])
        global clan_war_time
        date = clan_war_time
    else:
        track = input[13].split(',')
        select_num = input[15].split(',')
        for i in range(12):
            sheet.cell(3 + i, 12, track[i])
            for j in range(len(select_num)):
                if i + 1 == int(select_num[j]):
                    cell = sheet.cell(row=3 + i, column=12).coordinate
                    sheet[cell].font = Font(color='00B0F0', size=26, name='Cataneo BT')
        date = input[17]

    if int(input[7]) == 1:
        send_str = f"vs {input[5]} - {date}"
    else:
        send_str = f"vs {input[5]} ({input[7]}回目) - {date}"

    book.save(created_excel_pass)
    book.close()

    excel = win32com.client.Dispatch("Excel.Application")
    file = excel.Workbooks.Open(os.path.abspath(created_excel_pass))
    file.WorkSheets('6 vs 6').Select()
    file.ActiveSheet.ExportAsFixedFormat(0, os.path.abspath(pdf_pass))
    excel.Quit()

    pdf = convert_from_path(os.path.abspath(pdf_pass))  # pdfを画像化
    pdf[0].save(table_image_pass, "png")

    img = cv2.imread(table_image_pass)  # 余白を削除
    height, width = img.shape[:2]
    s = 1000
    h1, h2, w1, w2 = 0, 0, 0, 0

    for y in range(height):
        if all(img[y, s] != np.array([255, 255, 255])):
            h1 = y + 3
            break

    for y in reversed(range(height)):
        if all(img[y, s] != np.array([255, 255, 255])):
            h2 = y
            break

    for x in range(width):
        if all(img[s, x] != np.array([255, 255, 255])):
            w1 = x + 3
            break

    for x in reversed(range(width)):
        if all(img[s, x] != np.array([255, 255, 255])):
            w2 = x
            break

    crop_img = img[h1:h2, w1:w2]
    height, width = crop_img.shape[:2]
    crop_img = cv2.resize(crop_img, (int(width * 0.66), int(height * 0.66)))

    alpha_img = cv2.cvtColor(crop_img, cv2.COLOR_RGB2RGBA)  # 集計表の黒色を透過
    alpha_img[..., 3] = np.where(np.all(crop_img == 0, axis=-1), 0, 255)
    cv2.imwrite(table_image_pass, alpha_img)

    img1 = Image.open(bg_image_pass)  # 背景画像と集計表を合成
    img2 = Image.open(table_image_pass)
    img1.paste(img2, (0, 0), img2)
    img1.save(created_image_pass)

    os.remove(created_excel_pass)
    os.remove(pdf_pass)
    os.remove(table_image_pass)

    return send_str

File: Backup/Ver.2/make_table.rb
require 'rubyXL'
require 'win32ole'

$flag_war_start = false
$import_track = []


def maketable_introduction(flag)
  msg = "集計表を作成するお＾ｑ＾(Cancelで作成中止)\nコピペ用テンプレート↓\n"
  msg += "```自チームのメンバー [ふつきん,P2,P3,P4,P5,P6]\n自チームの各得点 [80,80,80,80,80,80]\n\n"
  msg += "敵チーム名 [ABC]\n対戦回数 [1]\n"
  msg += "敵チームのメンバー [P7,P8,P9,P10,P11,P12]\n敵チームの各得点 [80,80,80,80,80,80]"

  if flag == 1
    msg += "```※コースと交流戦日付は記入不要だお＾ｑ＾"
  else
    msg += "\n\nコース名 [T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12]\n"
    msg += "自チームの選択コース [1,2,3,4,5,6]\n\n日付 [2020/01/01]```"
  end

  return msg
end


def maketable_check_input(str, flag)
  safe, msg = true, ""
  n1 = str.count('[')
  n2 = str.count(']')

  if !str.start_with?("自チームのメンバー")
    safe = false
    msg = "Cancelと入力で作成中止できるお＾ｑ＾"
  elsif n1 >= 6 and n2 >= 6
    input = str.gsub("[", "]").split("]")
    1.step(11, 2) do |i|
      unless (i == 5 or i == 7) or input[i].count(',') == 5
        safe = false
        msg = "名前または得点が6つずつ入力されていないかも？＾ｑ＾"
      end
    end
    if flag == 0 and safe == true
      safe = false unless input[13].count(',') == 11
      msg = "コースが12個入力されていないかも？＾ｑ＾"
    end
    if flag == 1 and safe == true
      safe = false unless n1 == 6 and n2 == 6
      msg = "余計なデータがある or []で閉じられていない箇所があるかも？＾ｑ＾"
    elsif flag == 0 and safe == true
      safe = false unless n1 == 9 and n2 == 9
      msg = "余計なデータがある or []で閉じられていない箇所があるかも？＾ｑ＾"
    end
  else
    safe = false
    msg = "[]で閉じられていない箇所があるかも？＾ｑ＾"
  end

  return safe, msg
end


def make_table(input, flag)
  book = RubyXL::Parser.parse('D:\Ren\Desktop\色々\Data\Discord bot\Clan_War_table.xlsx')
  book.calc_pr.full_calc_on_load = true
  book.calc_pr.calc_completed = true
  book.calc_pr.calc_on_save = true
  book.calc_pr.force_full_calc = true
  sheet = book['6 vs 6']

  ally_name = input[1].split(',')
  6.times{|i| sheet[2 + i][3].change_contents(ally_name[i])}

  ally_point = input[3].split(',')
  6.times{|i| sheet[2 + i][5].change_contents(ally_point[i].to_i)}

  sheet[8][0].change_contents(input[5])

  enemy_name = input[9].split(',')
  6.times{|i| sheet[8 + i][3].change_contents(enemy_name[i])}

  enemy_point = input[11].split(',')
  6.times{|i| sheet[8 + i][5].change_contents(enemy_point[i].to_i)}

  if flag == 1
    12.times do |i|
      if $import_track[i].include?('__')
        sheet[2 + i][11].change_contents($import_track[i].gsub(/[^0-9a-z]/i, ""))
        sheet[2 + i][11].change_font_color('00B0F0')
      else
        sheet[2 + i][11].change_contents($import_track[i])
      end
    end
    date = $clan_war_time
  else
    track = input[13].split(',')
    select_num = input[15].split(',')
    j = 0
    12.times do |i|
      sheet[2 + i][11].change_contents(track[i])
      if i + 1 == select_num[j].to_i
        sheet[2 + i][11].change_font_color('00B0F0')
        j += 1
      end
    end
    date = input[17]
  end

  if input[7].to_i == 1
    send_str = "vs #{input[5]} - #{date}"
  else
    send_str = "vs #{input[5]} (#{input[7]}回目) - #{date}"
  end

  book.write('created_sheet.xlsx')


  excel = WIN32OLE.new('Excel.Application')
  book = excel.Workbooks.Open('D:\Ren\Desktop\色々\Data\Discord bot\created_sheet.xlsx')
  book.ExportAsFixedFormat({
        "FileName" => 'D:\Ren\Desktop\色々\Data\Discord bot\created_sheet.pdf',
        "Type" => 0,
        "OpenAfterPublish" => false
      })
  book.Close
  excel.Quit

  system('python convertPDFtoPNG.py')
  File.delete('created_sheet.xlsx')
  File.delete('created_sheet.pdf')

  return send_str
end
